lib/Mojolicious/Plugin/Authentication.pm

```diff
@@ -123,36 +117,137 @@ sub register {
         my $c = shift;
         return defined $current_user->($c);
     });
+    $app->helper(is_user_authenticated_p => sub {
+        my $c = shift;
+        $current_user_p->($c)->then(sub {
+            return defined $_[0];
+        });
+    });
 
     $app->helper($current_user_fn => $current_user);
+    $app->helper($current_user_fn."_p" => $current_user_p);
 
     $app->helper(logout => sub {
         my $c = shift;
         delete $c->stash->{$our_stash_key};
         delete $c->session->{$session_key};
         return 1;
     });
+}
 
-    $app->helper(authenticate => sub {
-        my ($c, $user, $pass, $extradata) = @_;
+# Unconditionally load the user based on uid in session
+sub user_loader_closure {
```

Comment:
I don't mind having the closures here instead of inline, and you've done a lot to reduce code repetition here, so that's great.

But it feels like maybe these two could be combined? I'm tempted by something like
``` perl
sub user_loader_closure {
    my ($stash_key, $session_key, $callback, $async) = @_;
    my $closure = sub {
        my $c = shift;
        my $uid = $c->session($session_key);
        return Mojo::Promise->resolve if !defined $uid;
        $callback->($c, $uid)->then(sub {
            my $user = $_[0];
            $c->stash(
                $stash_key => $user ? { user => $user } : { no_user => 1 }
            );
        });
    };
    return $closure if $async;
    # Not sure if this is the most idiomatic way to get a promise's value
    return sub {
        my @results;
        $closure->(@_)->then( sub { @results = @_ } )->wait;
    };
}
```
which we'd call only with the `_p` function as callback, and pass an extra value (in `$async`) for whether we want the promise at the end or not.

But I wonder if that might just introduce a needless performance hit for non-async path.

What do you think?

Reply:
I think it would be the needless performance hit. I did consider how to share code between the closure types, and as structured now it wouldn't be worth it. My thought is to deprecate the sync approach completely, since my promisification code will lightly wrap the sync code during deprecation period - and really, how many users does the module have?

Reply:
I had a go at a "depromisify" function. However, it's impossible because the event loop is running. The only way to achieve making the synchronous version work with promise code would be with async/await. Therefore, I submit that this is the only way to make it work.

Reply:
The closest we could get, at a minor performance cost, would be to split the three functions (load, validate, authenticate) each into three parts: prepare, execute (the user-supplied bit), and process the results. The first and last bits would be the same both sync and async, and they'd be fed into a generalised sync-and-async wrapper-maker. Because the prepare and results-processing are so small though, you'd be adding overhead to a fairly hot bit of code.

Reply:
Hm, yeah, I see what you mean. I think this can be merged as is then, and if async / await becomes a possibility, or if the need arises, we can tweak this further to avoid the repetition. I see no reason to block this only on those grounds.

Thanks for all the work you've put into this!